### PR TITLE
Clean-up Python #4815 - P2.2: Chem\Descriptors

### PR DESCRIPTION
**Pull request by IchiruTake**
This is the sub-PR breaking from the large PR 4815: https://github.com/rdkit/rdkit/pull/4815

-----
Updated Changes:
- [Bug Fix]: Optimize function in `GraphDescriptors.py` required another class `Counter` in `collections`. This caused error in `Descriptors.py`. this patch is to ensure that all descriptors can be append if belong to the RDKit
- [Opt]: Optimize several `GraphDescriptors.py`
- [New]: Function `_ChargeDescriptors()` is faster which linearly proportional to the molecule

### CODE REVIEW

File: rdkit/Chem/Descriptors.py [hunk 1]
```diff
@@ -7,28 +7,51 @@
...
+  
+  try:
+    if something.__class__.__module__ == NAME:
+      return True
+    if NAME in something.__class__.__module__:
+      return True
+  except Exception:
+    pass
+  
+  return False
```
Comment (gedeck) [inline rdkit/Chem/Descriptors.py:46]: This is a significant modification. Why is this required and what is it filtering out?

File: rdkit/Chem/Descriptors.py [hunk 2]
```diff
@@ -149,31 +172,34 @@ def NumRadicalElectrons(mol):
...
+      maxChg = chg
   res = (minChg, maxChg)
   mol._chargeDescriptors = res
   return res
 
 
 def MaxPartialCharge(mol, force=False):
-  _, res = _ChargeDescriptors(mol, force)
-  return res
+  return _ChargeDescriptors(mol, force)[1]
```
Comment (gedeck) [inline rdkit/Chem/Descriptors.py:195]: I'm not sure if this is clearer. A namedtuple would be beneficial.
```
from collections import namedtuple

ChargeDescriptors = namedtuple('ChargeDescriptors', 'minCharge, maxCharge')

def _ChargeDescriptors(mol, force=False):
  """ Returns the charge descriptions of the molecule in a specific range: 2-value tuple
  
  """
  if not force and hasattr(mol, '_chargeDescriptors'):
    return mol._chargeDescriptors
  rdPartialCharges.ComputeGasteigerCharges(mol)
  minChg = 500.
  maxChg = -500.
  for at in mol.GetAtoms():
    chg = float(at.GetProp('_GasteigerCharge'))
    if chg < minChg:
      minChg = chg
    elif chg > maxChg:
      maxChg = chg
  res = ChargeDescriptors(minChg, maxChg)
  mol._chargeDescriptors = res
  return res

def MaxPartialCharge(mol, force=False):
  return _ChargeDescriptors(mol, force).minCharge
```

File: rdkit/Chem/GraphDescriptors.py [hunk 3]
```diff
@@ -34,28 +33,22 @@ def _VertexDegrees(mat, onlyOnes=0):
...
 
 def _GetCountDict(arr):
   """  *Internal Use Only*
 
   """
-  res = {}
-  for v in arr:
-    res[v] = res.get(v, 0) + 1
-  return res
+  return dict(Counter(arr))
```
Comment (gedeck) [inline rdkit/Chem/GraphDescriptors.py:51]: I don't think this will be any faster.

File: rdkit/Chem/Descriptors.py [hunk 4]
```diff
@@ -7,28 +7,51 @@
 #  which is included in the file license.txt, found at the root
 #  of the RDKit source tree.
 #
-from collections import abc  # this won't work in python2, but we don't support that any more
 
+from collections import abc  # This won't work in python 2, but we don't support that any more
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/Descriptors.py:11]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/Descriptors.py [hunk 5]
```diff
@@ -7,28 +7,51 @@
...
 from rdkit.Chem.QED import qed
 
 
-def _isCallable(thing):
-  return isinstance(thing, abc.Callable) or \
-              hasattr(thing, '__call__')
-
-
+def _isCallable(something):
+  return isinstance(something, (abc.Callable, Callable)) or hasattr(something, '__call__')
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/Descriptors.py:25]: I don't really know the hidden implementation, but I expect some pre-defined descriptors have this implementation, so I did not make any change here. You can see it from the orginial source code, I guess

File: rdkit/Chem/Descriptors.py (hunk 1 above)
Comment (IchiruTake) [reply rdkit/Chem/Descriptors.py:46]: Yes, just for notice here, if some descriptors used any third-party modules or libraries, we want to ensure those should be consistent called, I have made so many mistake on this issue before getting worked done. This is to guarantee and prevent further issue that can be potentially made. 
Moreover, `Descriptors` is easily to be made as the Chemistry development, it would be noticed that this issue can be pointed out in the early version

File: rdkit/Chem/Descriptors.py [hunk 6]
```diff
@@ -7,28 +7,51 @@
...
+  
+  return False
+  
 _descList = []
 
 
 def _setupDescriptors(namespace):
   global _descList, descList
-  from rdkit.Chem import GraphDescriptors, MolSurf, Lipinski, Fragments, Crippen, Descriptors3D
+  from rdkit.Chem import Descriptors3D  # Shortcut
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/Descriptors.py:53]: This is just an issue automatic I have made to point out unused import, but I forgot to turn it off in this PR, I will revert it back

File: rdkit/Chem/Descriptors.py [hunk 7]
```diff
@@ -37,7 +60,7 @@ def _setupDescriptors(namespace):
   otherMods = [Chem]
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/Descriptors.py:60]: Yes, in the new function, it do. But to be honest, I wonder why `Descriptors3D` is not added. What do you think @greglandrum? Should I add it or just leave there for discussion, I saw `Chem` is added into `otherMods`. I am not confident with this change

File: rdkit/Chem/Descriptors.py (hunk 2 above)
Comment (IchiruTake) [reply rdkit/Chem/Descriptors.py:195]: [FYI] Yes, I agree with this change, but if (just if) making this change in other module will crash it (similar to the commit made on `GraphDescriptors.py`

Comment (IchiruTake) [reply rdkit/Chem/Descriptors.py:195]: I am afraid that this could break the compatibility, I would prefer this `ChargeDescriptors = namedtuple('ChargeDescriptors', ['minCharge', 'maxCharge'])`. This allowed that `ChargeDescriptors[0] = ChargeDescriptors.minCharge` and similar for `maxCharge`

File: rdkit/Chem/GraphDescriptors.py (hunk 3 above)
Comment (IchiruTake) [reply rdkit/Chem/GraphDescriptors.py:51]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

Comment (IchiruTake) [reply rdkit/Chem/GraphDescriptors.py:51]: [FYI]: This function is not called from anywhere, so I just convert it to `Counter(arr)`. The reason I still keep this change is also help that adding third-party library is safe to do so at any other modules apart from `Descriptors.py`.

File: rdkit/Chem/GraphDescriptors.py [hunk 8]
```diff
@@ -216,12 +200,11 @@ def Chi1(mol):
   """ From equations (1),(11) and (12) of Rev. Comp. Chem. vol 2, 367-422, (1991)
 
   """
-  c1s = [x.GetBeginAtom().GetDegree() * x.GetEndAtom().GetDegree() for x in mol.GetBonds()]
-  while 0 in c1s:
-    c1s.remove(0)
+  c1s = [x.GetBeginAtom().GetDegree() * x.GetEndAtom().GetDegree() 
+         for x in mol.GetBonds() 
+         if x.GetBeginAtom().GetDegree() != 0 and x.GetEndAtom().GetDegree() != 0]
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/GraphDescriptors.py:205]: Yes, but I prefer normal `for` loop rather making an temporary list to filter out

File: rdkit/Chem/GraphDescriptors.py [hunk 9]
```diff
@@ -255,16 +238,15 @@ def _hkDeltas(mol, skipHs=1):
...
   """
   deltas = _hkDeltas(mol)
-  while 0 in deltas:
-    deltas.remove(0)
+  deltas_ = [x for x in deltas if x != 0.0 or x != 0 or x != -0.0]
+  del deltas
   mol._hkDeltas = None
-  res = sum(numpy.sqrt(1. / numpy.array(deltas)))
+  res = sum(numpy.sqrt(1. / numpy.array(deltas_)))
   return res
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/GraphDescriptors.py:250]: Agree, I will fix it in the next commit

File: rdkit/Chem/GraphDescriptors.py [hunk 10]
```diff
@@ -330,24 +312,28 @@ def _pyChi0n(mol):
...
 
   """
-  deltas = [_nVal(x) for x in mol.GetAtoms()]
-  while deltas.count(0):
-    deltas.remove(0)
+  deltas = []
+  for x in mol.GetAtoms():
+    value = _nVal(x)
+    if value != 0:
+      deltas.append(value)
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/GraphDescriptors.py:319]: I don't understand why should it needed to loop twice, while I can make a check for that

File: rdkit/Chem/GraphDescriptors.py [hunk 11]
```diff
@@ -330,24 +312,28 @@ def _pyChi0n(mol):
...
   res = 0.0
   for bond in mol.GetBonds():
-    v = delts[bond.GetBeginAtomIdx()] * delts[bond.GetEndAtomIdx()]
-    if v != 0.0:
-      res += numpy.sqrt(1. / v)
+    v = deltas[bond.GetBeginAtomIdx()] * deltas[bond.GetEndAtomIdx()]
+    res += numpy.sqrt(1. / v)
   return res
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/GraphDescriptors.py:337]: Agree ! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/UnitTestDescriptors.py [hunk 12]
```diff
@@ -11,58 +11,73 @@
...
         except RuntimeError:
           # 3D descriptors fail since the mol has no conformers
           pass
         except Exception:
           import traceback
           traceback.print_exc()
+          # __ShowDescriptorErrors(index=i, name=nm, function=fn)
           raise AssertionError('SMILES: %s; Descriptor: %s' % (smi, nm))
 
+  def testInvalidDescriptors(self):
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/UnitTestDescriptors.py]: Yes, you got my point

File: rdkit/Chem/Descriptors.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gedeck) [reply rdkit/Chem/Descriptors.py:25]: I was referring to `abc.Callable` being changed to `abc.Callable, Callable`

File: rdkit/Chem/GraphDescriptors.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (gedeck) [reply rdkit/Chem/GraphDescriptors.py:321]: Either is fine as long it's not the original `while` loop. That looks too inefficient. Going through the list repeatedly instead of having only one pass.

File: rdkit/Chem/GraphDescriptors.py [hunk 13]
```diff
@@ -330,24 +315,23 @@ def _pyChi0n(mol):
...
   """
-  delts = numpy.array([_nVal(x) for x in mol.GetAtoms()], 'd')
+  deltas = [_nVal(atom) for atom in mol.GetAtoms()]
   res = 0.0
   for bond in mol.GetBonds():
-    v = delts[bond.GetBeginAtomIdx()] * delts[bond.GetEndAtomIdx()]
-    if v != 0.0:
-      res += numpy.sqrt(1. / v)
+    v = deltas[bond.GetBeginAtomIdx()] * deltas[bond.GetEndAtomIdx()]
+    res += numpy.sqrt(1. / v)
```
Comment (gedeck) [inline rdkit/Chem/GraphDescriptors.py:334]: This is missing the check for `v != 0`

**Review by gedeck — CHANGES_REQUESTED** [review]
One question still open, and one refactor missing an `if` statement
